harvester/DBInterface.py

```diff
@@ -267,41 +225,171 @@ def delete_record(self, record):
+				self.delete_related_records("records_x_access", record['record_id'])
+				self.delete_related_records("records_x_creators", record['record_id'])
+				self.delete_related_records("records_x_publishers", record['record_id'])
+				self.delete_related_records("records_x_rights", record['record_id'])
+				self.delete_related_records("records_x_subjects", record['record_id'])
+				self.delete_related_records("records_x_tags", record['record_id'])
+				self.delete_related_records("descriptions", record['record_id'])
+				self.delete_related_records("geospatial", record['record_id'])
+				self.delete_related_records("domain_metadata", record['record_id'])
 			except:
 				self.logger.error("Unable to delete related table rows for record %s" % (record['local_identifier'],))
 				return False
 
 		self.logger.debug("Marked as deleted: record %s" % (record['local_identifier'],))
 		return True
 
-	def get_record_id(self, repo_id, local_identifier):
-		returnvalue = None
+	def delete_related_records(self, crosstable, record_id):
 		con = self.getConnection()
 		with con:
 			cur = self.getCursor(con)
-			cur.execute(self._prep("select record_id from records where local_identifier=? and repository_id = ?"),
-						(local_identifier, repo_id))
+			try:
+				sqlstring1 = "DELETE from {} where record_id=?".format(crosstable)
+				cur.execute(self._prep(sqlstring1), (record_id,))
+			except:
+				return False
+
+		return True
+
+	def get_table_id_column(self, tablename):
+		if tablename == "rights":
+			return "rights_id"
+		elif tablename == "access":
+			return "access_id"
+		elif tablename == "repositories":
+			return "repository_id"
+		elif tablename == "geospatial":
+			return "geospatial_id"
+		elif tablename == "domain_metadata":
+			return "metadata_id"
+		elif tablename == "domain_schemas":
+			return "schema_id"
+		elif "records_x_" in tablename:
+			return tablename + "_id"
+		else:
+			return tablename[:-1] + "_id"
+
+	def get_table_value_column(self, tablename):
+		if tablename == "rights":
+			return "rights_hash"
+		elif tablename == "access":
+			return "access"
+		elif tablename == "geospatial":
+			return "coordinate_type"
+		elif tablename == "domain_metadata":
+			return "schema_id"
+		elif tablename == "domain_schemas":
+			return "namespace"
+		elif tablename == "repositories":
+			return "repository_url"
+		elif tablename == "records":
+			return "local_identifier"
+		else:
+			return tablename[:-1]
+
+	def insert_related_record(self, tablename, val, **kwargs):
+		valcolumn = self.get_table_value_column(tablename)
+		idcolumn = self.get_table_id_column(tablename)
+		related_record_id = None
+		paramlist = [val]
+		sqlstring1 = "INSERT INTO {} ({}".format(tablename, valcolumn)
+		sqlstring2 = "VALUES(?"
+		for key, value in kwargs.items():
+			sqlstring1 += "," + str(key)
+			sqlstring2 += ",?"
+			paramlist.append(value)
+		sqlstring1 += ")"
+		sqlstring2 += ")"
+		con = self.getConnection()
```

Comment:
This is some really potentially yucky string concatentation ... I might feel better if they weren't just called `sqlstring1` and `sqlstring2` (especially since as far as I can tell they're always just whitespace separated?)

Reply:
I agree, but the DB drivers do not allow for "?" placeholders in the table names or table column names.  Took me an hour to figure that problem out.  My previous solution was elegant but wrong.